service/storage/auxiliary_test.go

```diff
@@ -0,0 +1,299 @@
+	db := helpers.InMemoryDB(t)
+	defer db.Close()
+	txn := db.NewTransaction(false)
+
+	// This is a success func that is never expected to be called.
+	noCallFn := func(txn *badger.Txn) error {
+		t.Fail()
+		return nil
+	}
+	successFn := func(txn *badger.Txn) error {
+		return nil
+	}
+	failFn := func(txn *badger.Txn) error {
+		return errors.New("fail")
+	}
+
+	t.Run("should error if all ops failed", func(t *testing.T) {
+		err := Fallback(
+			failFn,
+			failFn,
+			failFn,
+			failFn,
+		)(txn)
+
+		assert.Error(t, err)
+		merr, ok := err.(*multierror.Error)
+		assert.True(t, ok)
+		assert.Len(t, merr.Errors, 4)
+	})
+
+	t.Run("should not error if fallback succeeds", func(t *testing.T) {
+		err := Fallback(
+			failFn,
+			successFn,
+			noCallFn,
+			noCallFn,
+		)(txn)
+
+		assert.NoError(t, err)
+	})
+
+	t.Run("should not error if any fallback succeeds", func(t *testing.T) {
+		err := Fallback(
+			failFn,
+			failFn,
+			failFn,
+			failFn,
+			failFn,
+			successFn,
+			noCallFn,
+			noCallFn,
+			noCallFn,
+		)(txn)
+
+		assert.NoError(t, err)
+	})
+
+	t.Run("should not call second fallback op if first fallback succeeds", func(t *testing.T) {
+		err := Fallback(
+			failFn,
+			failFn,
+			failFn,
+			successFn,
+			noCallFn,
+			noCallFn,
+			noCallFn,
+		)(txn)
+
+		assert.NoError(t, err)
+	})
+
+	t.Run("should not call fallback if first op succeeds", func(t *testing.T) {
+		err := Fallback(
+			successFn,
+			noCallFn,
+			noCallFn,
+			noCallFn,
+			noCallFn,
+		)(txn)
+
+		assert.NoError(t, err)
+	})
+}
+
+func TestCombine(t *testing.T) {
+	db := helpers.InMemoryDB(t)
+	defer db.Close()
+	txn := db.NewTransaction(false)
+
+	// This is a success func that is never expected to be called.
+	noCallFn := func(txn *badger.Txn) error {
+		t.Fail()
+		return nil
+	}
+	successFn := func(txn *badger.Txn) error {
+		return nil
+	}
+	failFn := func(txn *badger.Txn) error {
+		return errors.New("fail")
+	}
+
+	t.Run("nominal case", func(t *testing.T) {
+		calls := 0
+		f := func(txn *badger.Txn) error {
+			calls++
+			return nil
+		}
+		err := Combine(f, f, f, f, f, f)(txn)
+
+		assert.NoError(t, err)
+		assert.Equal(t, calls, 6)
+	})
+
+	t.Run("should error if first op fails", func(t *testing.T) {
+		err := Combine(
+			failFn,
+			noCallFn,
+			noCallFn,
+			noCallFn,
+		)(txn)
+
+		assert.Error(t, err)
+	})
+
+	t.Run("should error if last op fails", func(t *testing.T) {
+		err := Combine(
+			successFn,
+			failFn,
+		)(txn)
+
+		assert.Error(t, err)
+	})
+
+	t.Run("should error if any op fails", func(t *testing.T) {
+		err := Combine(
+			successFn,
+			successFn,
+			successFn,
+			successFn,
+			failFn,
+			noCallFn,
+			noCallFn,
+		)(txn)
+
+		assert.Error(t, err)
+	})
+}
+
+func TestRetrieve(t *testing.T) {
+	db := helpers.InMemoryDB(t)
+	defer db.Close()
+
+	// Insert test value.
+	const testValue = uint64(42)
+	testKey := []byte{42}
+
+	t.Run("nominal case", func(t *testing.T) {
+		insertKeyValue(t, db, testKey, testValue)
```

Comment:
I generally prefer to have error returns and check them with `require.NoError(err)`, then to pass on the testing argument to the function. It just seems more flexible. What do you think?